File: examples/operators/com.ibm.streamsx.testing.examples.operators.java/src/com/ibm/streamsx/testing/examples/operators/java/TestMultiplyBy.java
package com.ibm.streamsx.testing.examples.operators.java;

import java.io.File;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.TimeUnit;

import org.junit.Assert;
import org.junit.Test;

import com.ibm.streams.operator.StreamSchema;
import com.ibm.streams.operator.Type;
import com.ibm.streamsx.topology.TStream;
import com.ibm.streamsx.topology.Topology;
import com.ibm.streamsx.topology.context.StreamsContextFactory;
import com.ibm.streamsx.topology.spl.SPL;
import com.ibm.streamsx.topology.spl.SPLStream;
import com.ibm.streamsx.topology.tester.Condition;
import com.ibm.streamsx.topology.tester.Tester;


/*
 * This class hows how one can test a SPL composite operator using JUnit
 * and the Streams Java Application API.
 */
public class TestMultiplyBy {
	
	/*
	 * Method to create the topology, submit the application based on 
	 * the specified context and validate data
	 */
	private void doTestMultiplyBy(String context) throws Exception {
		
		// Create topology to drvie the test
		Topology topo = new Topology("TestMultiplyBy");
				
		// Add the toolkit under test to the toolkit path
		SPL.addToolkit(topo, new File("../com.ibm.streamsx.testing.examples.operators.spl"));
		SPL.addToolkit(topo, new File("../com.ibm.streamsx.testing.examples.operators.app"));
		
		// Get SPL Schema for test result
		StreamSchema schema = Type.Factory.getStreamSchema("tuple<int32 result>");
		
		// Set up parameters for the composite operator
		Map<String, Object> params = new HashMap<String, Object>();
		params.put("factor", 3);
		
		// Invoke the test composite
		SPLStream resultStream = SPL.invokeSource(topo, "com.ibm.streamsx.testing.examples.operators.spl::TestMultiplyBy", params, schema);
		
		// Convert the SPL Stream to TStream so we can work with it with the tester
		TStream<Integer> tStream = resultStream.convert(t->{
			return t.getInt("result");
		});
				
		// Set up tester to validate test result
		Tester tester = topo.getTester();
		Condition tupleCnt = tester.tupleCount(tStream, 34);
		
		// Submit the test.  Test will run and timeout in 10 seconds
		tester.complete(StreamsContextFactory.getStreamsContext(context), tupleCnt, 10, TimeUnit.SECONDS);
		
		// Validate that the test generate correct tuple count
		Assert.assertTrue(tupleCnt.valid());	
	}
	
	
	/*
	 * Test in standalone mode
	 */
	@Test
	public void testStandalone() throws Exception {		
		doTestMultiplyBy("STANDALONE_TESTER");
		
	}
	
	/*
	 * Test in distributed mode
	 */
	@Test
	public void testDistributed() throws Exception {
		 doTestMultiplyBy("DISTRIBUTED_TESTER");
	}

}

File: examples/operators/com.ibm.streamsx.testing.examples.operators.python/testMultiplyBy.py
import unittest

from streamsx.topology.topology import *
from streamsx.topology.tester import Tester
import streamsx.spl.op as op
import streamsx.spl.toolkit as tk
import numpy as np

class TestMultiplyByStandalone(unittest.TestCase):
    """TestCustomStandalone class for driving the test in standalone mode"""
    def setUp(self):
        ''' Set up to run test as Standalone'''
        Tester.setup_standalone(self)
    def _add_toolkits(self, topo):
        ''' Add required toolkits for test to run '''
        tk.add_toolkit(topo, '../com.ibm.streamsx.testing.examples.operators.app')
        tk.add_toolkit(topo, '../com.ibm.streamsx.testing.examples.operators.spl')
    def test_op(self):
        ''' Create topology to drive the test from com.ibm.streamsx.testing.spl '''
        topo = Topology()
        self._add_toolkits(topo)

        '''Set up parameter to call the test composite'''
        params = {'factor':3}

        ''' Call the test composite'''
        test_op = op.Source(topo, 'com.ibm.streamsx.testing.examples.operators.spl::TestMultiplyBy', 'tuple<int32 result>', params=params)

        ''' Convert the SPLStream to Python Stream so we can work with the data in the tester '''
        mapped = test_op.stream.map(lambda x: x['result'])

        ''' Set up Tester to validate the result of running the test composite'''
        tester = Tester(topo)

        ''' Example to check for tuple count'''
        tester.tuple_count(test_op.stream, 34)

        ''' Example to check content of the stream with an expected list of data'''
        expected = list(np.arange(0, 100, 3))
        tester.contents(mapped, expected)

        ''' Example to check data tuple by tuple'''
        tester.tuple_check(test_op.stream, lambda  x:(x['result']%3)==0)

        tester.test(self.test_ctxtype, self.test_config)


class TestMultiplyByDistributed(TestMultiplyByStandalone):
    """Example to run the same test in distributed mode"""
    def setUp(self):
        Tester.setup_distributed(self)


#class TestFTPCloud(TestFTP):
#    """ Test invocations of FTP operators with streaming analytics """
#    def setUp(self):
#        Tester.setup_streaming_analytics(self, force_remote_build=True)
